### PR TITLE
Fix some issues with elevation profile vector layer results

### LINKED ISSUE #49279: ZPolyLine Profile with new profile tool doesn't work
### What is the bug or the crash?

The Profile widget (new in 3.26).
When using a ZPolyLine to "Capture Curve for Feature", the ZPolyLine itself is not displayed on the profile.

### Steps to reproduce the issue

I am trying out the new QGIS 3.26 profile tool. I imported a SRTM DEM. I can create a ZPolyLine and use it to plot the Terrain Elevation of the SRTM layer under the line by:

   1. Ticking the SRTM Layer in the profile widget,
    2. Selecting the "Capture Curve from Feature" tool and selecting my ZPolyLine in the map. all good so far.

Now I want to add the ZPolyLine itself (The Z values of the ZPolyline are slightly higher that the corresponding SRTM values) to the profile. I would expect to just have to tick the ZPolyLine layer in the profile tool. But that doesn't display it.

### Versions

Linux 3.26
Linux build of today's master: 2f3d6b, so 3.27

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

The workaround is to extract the vertices from the ZPolyLine. Ticking the layer with the extracted vertices displays the profile of the vertices, and making it a "Surface" brings the line between the vertices. but there must be something easier right?
There is a question open for the topic: https://gis.stackexchange.com/q/435350?sem=2

### LINKED ISSUE #52815: Elevation Profile: intersections with MultiPolygonZ sometimes generate artifacts and gaps
### What is the bug or the crash?

When intersecting an elevation profile line with a dataset representing buildings (MultiPolygonZ geometry) there are often gaps and sometimes artifacts (triangles into "infinity").

image

### Steps to reproduce the issue

1. Download the data from [https://www.carto.net/neumann/temp/profile_issues.zip](https://www.carto.net/neumann/temp/profile_issues.zip)
2. Unzip the data
3. Open the project "profile_issues.qgs"
4. Open the elevation profile panel
5. From the elevation profile use the "Capture Curve from Feature tool" 
image
 and select in the 2D map view the purple profile Line in the "profile_lines" layer that runs north-south
6. Observe the profile with the issues (gaps and artefacts)

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.31.0-Master | QGIS code revision | 0ce567bd306
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.6
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.2 LTS
  |   |   |  
Active Python plugins
changeDataSource | 3.1
solocator | 1.6.0
pg_raster_import | 3.2.0
trackable_project_files | v1.2.2
linking_relation_editor | v1.1.9
QgisModelBaker | v7.5.0
AtlasExportFunction | 1.1
processing | 2.12.99
db_manager | 0.1.20

</body></html>QGIS version
3.31.0-Master
QGIS code revision
[0ce567bd306](https://github.com/qgis/QGIS/commit/0ce567bd306)
Qt version
5.15.3
Python version
3.10.6
GDAL/OGR version
3.4.1
PROJ version
8.2.1
EPSG Registry database version
v10.041 (2021-12-03)
GEOS version
3.10.2-CAPI-1.16.0
SQLite version
3.37.2
PDAL version
2.3.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.4
QScintilla2 version
2.11.6
OS version
Ubuntu 22.04.2 LTS

Active Python plugins
changeDataSource
3.1
solocator
1.6.0
pg_raster_import
3.2.0
trackable_project_files
v1.2.2
linking_relation_editor
v1.1.9
QgisModelBaker
v7.5.0
AtlasExportFunction
1.1
processing
2.12.99
db_manager
0.1.20

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nyalldawson**
Fixes https://github.com/qgis/QGIS/issues/52815
Fixes https://github.com/qgis/QGIS/issues/49279

### CODE REVIEW

File: src/core/geometry/qgsgeometryutils.h [hunk 1]
```diff
@@ -789,6 +789,16 @@ class CORE_EXPORT QgsGeometryUtils
      */
     static double triangleArea( double aX, double aY, double bX, double bY, double cX, double cY ) SIP_HOLDGIL;
 
+    /**
+     * Given the line (\a x1, \a y1) to (\a x2, \a y2) and a point (\a px, \a py) returns the percentage
```
Comment (nicogodet) [inline src/core/geometry/qgsgeometryutils.h:793]:
```suggestion
     * Given the line (\a x1, \a y1) to (\a x2, \a y2) and a point (\a px, \a py) returns the fraction
```

percentage would imply a return value between 0 and 100

File: src/core/geometry/qgsgeometryutils.h [hunk 2]
```diff
@@ -809,6 +819,14 @@ class CORE_EXPORT QgsGeometryUtils
     static void weightedPointInTriangle( double aX, double aY, double bX, double bY, double cX, double cY,
                                          double weightB, double weightC, double &pointX SIP_OUT, double &pointY SIP_OUT ) SIP_HOLDGIL;
 
+    /**
+     * Given the points (\a x1, \a y1), (\a x2, \a y2) and (\a x3, \a y3) returns TRUE if these
+     * points can be considered collinear with a specified tolerance.
```
Comment (nicogodet) [inline src/core/geometry/qgsgeometryutils.h:824]:
```suggestion
     * points can be considered collinear with a specified tolerance \a epsilon.
```

**Review by alexbruy — APPROVED** [review]

File: tests/src/core/geometry/testqgsgeometryutils.cpp [hunk 3]
```diff
@@ -1376,7 +1378,27 @@ void TestQgsGeometryUtils::testPointOnLineWithDistance()
...
   QGSCOMPARENEAR( x, 8.57493, 0.0001 );
   QGSCOMPARENEAR( y, 5.14496, 0.0001 );
+}
 
+void TestQgsGeometryUtils::testPointFractionAlongLine()
+{
+  QGSCOMPARENEAR( QgsGeometryUtils::pointFractionAlongLine( 0, 10, 20, 30, 0, 10 ), 0, 0.00001 );
+  QGSCOMPARENEAR( QgsGeometryUtils::pointFractionAlongLine( 0, 10, 20, 30, 20, 30 ), 1.0, 0.00001 );
+  QGSCOMPARENEAR( QgsGeometryUtils::pointFractionAlongLine( 0, 10, 20, 10, 10, 10 ), 0.5, 0.00001 );
+}
```
Comment (MorriganR) [inline tests/src/core/geometry/testqgsgeometryutils.cpp:1388]:
```suggestion
  QGSCOMPARENEAR( QgsGeometryUtils::pointFractionAlongLine( 40000.0, 40000.00001, 40000.00002, 40000.00001, 40000.00001, 40000.00001 ), 0.5, 0.00001 );
}
```
add check for accuracy

File: src/core/geometry/qgsgeometryutils.h (hunk 2 above)
Comment (nyalldawson) [reply src/core/geometry/qgsgeometryutils.h:824]: Thanks, all addressed!

File: tests/src/core/geometry/testqgsgeometryutils.cpp (hunk 3 above)
Comment (nyalldawson) [reply tests/src/core/geometry/testqgsgeometryutils.cpp:1389]: Added, thanks!

**Pull request merged**